### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by justbur** [justbur/olig#1]
@TheBB @syl20bnr

I've been meaning to put something like this together for a while. This library gives a new way to think about "leader" keys. You define a leader, the keys it's triggered by, the modes it's activated by (minor or major), and any initial bindings. Here's a sketch of how it would work in spacemacs.

```elisp
(olig-define base-key
  :evil-keys ("SPC")                                            
  :keys ("M-m")                                                 
  :bindings                                                     
  "h" (lambda () (interactive) (message "hi from base")))
(olig-define mode-key
  :keys ("M-m m")                                               
  :evil-keys ("SPC m" ",")                            
  :evil-states (normal motion)          
  :active-when (emacs-lisp-mode)                                
  :bindings                                                     
  "h" (lambda () (interactive) (message "hi from mode")))

(olig-set-keys mode-key "b" (lambda () (interactive) (message "bye!")))
```

**Comment by justbur** [comment]
I didn't think much about the name yet. I just put something in that was simple to type

**Comment by justbur** [comment]
Added a `:default-bindings` keyword and a corresponding `olig-set-key-defaults`. Default bindings never override existing ones.

**Comment by syl20bnr** [comment]
For the name I officially write my (hopefully) best proposition: `star-key`
In the README we can document it as `STARting-key` so it fit the space semantic while conveying the idea of the leader _starting_ key of a sequence.

**Comment by syl20bnr** [comment]
Main downside: the star key is an actual key...

**Comment by syl20bnr** [comment]
So we could make it plural form `star-keys`

**Comment by TheBB** [comment]
Technically that key is an asterisk.

**Comment by syl20bnr** [comment]
oh right... good :-)

**Comment by justbur** [comment]
I'm not sure about `star-key`. Looking at it without knowing it's short for starting-key, I don't think I'd have any idea what it does.

**Comment by justbur** [comment]
Any thoughts on design?

**Comment by syl20bnr** [comment]
A good name first :-)

Seriously I have not looked at it yet, will do.

Le mercredi 11 novembre 2015, Justin Burkett <notifications@github.com> a
écrit :

> Any thoughts on design?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/justbur/olig/issues/1#issuecomment-155866992>.
>


-- 
-syl20bnr-

**Comment by justbur** [comment]
Was thinking it would be useful to put the lisp stuff on a leader key and free up SPC k

**Comment by justbur** [comment]
This is cool I think
```elisp
  (olig-define-leader lisp-leader
    :evil-keys ("K")
    :mode-maps (emacs-lisp-mode-map
                lisp-mode-map
                ;; any others
                )
    :init-map evil-lisp-state-map)
```

**Comment by justbur** [comment]
Also, it's possible to easily test this in spacemacs now (for the non-mode-based bindings) using
```elisp
  (global-evil-leader-mode -1)
  (olig-define-leader spacemacs-leader
    :keys ("M-m")
    :evil-keys ("SPC")
    :use-map evil-leader--default-map)
```

**Comment by justbur** [comment]
It's boring, but the name `bind-keymap` or just `bind-map` (maybe `bind-tree` or `key-tree`) conveys the right idea, because all that this does is allows you to bind a keymap in multiple places and to make it contingent on modes and evil states easily.

**Comment by syl20bnr** [comment]
Interesting direction you took, so the package will be a light abstraction to bind any map to some keys ?

**Comment by justbur** [comment]
Yes that's the idea. Each leader has a map, and is bound to one or more keys where the binding depends on the major or minor mode. 

Using a minor mode for this like evil-leader does is unnecessary and overly complicated imo

**Comment by syl20bnr** [comment]
I'm with you on this design, seems both simple and closer to the Emacs way of doing things.
`bind-map` is a really good name for this, cannot think about a better one.

**Comment by justbur** [comment]
Great thanks. I like bind-map too

**Comment by syl20bnr** [comment]
`bind-keymap` is more accurate though, `bind-map` does not tell what kind of map.

**Comment by syl20bnr** [comment]
I'm looking at `bind-key` package and maybe this functionality is already available, can you look for `:prefix-map` in this package ?

**Comment by justbur** [comment]
It probably doesn't interface with evil though

**Comment by syl20bnr** [comment]
Oh... but it is `keymap` driven, not `mode` driven, so `keymap` must exist. I believe.

**Comment by syl20bnr** [comment]
Right, maybe `bind-key` is a useful dependency for `bind-keymap`.

**Comment by justbur** [comment]
Maybe, I'll look. I can handle everything that bind-key would do with
define-key though. I don't think I need the extra features from that package

**Comment by syl20bnr** [comment]
I read some talk about bind-key being added to ELPA (at least, may go into core if use-package makes its way into it), I let you see if this abstraction does help `bind-(key)map` or not.

**Comment by justbur** [comment]
My understanding is that unless you are using bind-key* there is not much
difference from define-key

**Comment by justbur** [comment]
@syl20bnr the other important feature for spacemacs here is the default
bindings idea which never overwrite user bindings.

**Comment by syl20bnr** [comment]
We should not need this feature. The issue raised in https://github.com/syl20bnr/spacemacs/commit/830212c509b3d5888168e60faec64280f6e59e6d is a non issue, we should not care about it.
All we need is to be able to map keys for specific modes, then we are free to define autoloads like we want, this is a base Emacs feature. So don't be bothered by the non issue raised for flycheck config.

cc @TheBB

**Comment by justbur** [comment]
We don't need it but it will be helpful because then the user doesn't have
his or her bindings overwritten when a package is deferred. They also won't
have to worry about putting them in user-init or user-config too. It's not
the flycheck issue I'm referring to.

**Comment by syl20bnr** [comment]
As long as spacemacs defines them in `:init` I don't see any issue.
Of course the user must use the same function as the one used in Spacemacs.

Do you want a unique function to do all our key bindings need in Spacemacs?

**Comment by justbur** [comment]
> As long as spacemacs defines them in :init I don't see any issue.
Of course the user must use the same function as the one used in Spacemacs.

Assuming that it's done consistently throughout and users always understand to use `user-config` then yes the problem can be avoided, but I still see it as good practice, along the lines of why `defvar` and `defcustom` don't overwrite user settings. 

> Do you want a unique function to do all our key bindings need in Spacemacs?

I have one in here that can be used, but it's not necessary. Once the map is defined and bound in the right places, you're free to use what you want to add keys to that map.

**Comment by syl20bnr** [comment]
> Assuming that it's done consistently throughout and users always understand to use user-config

This is not an assumption, it is how spacemacs must be used :-)

>  I still see it as good practice, along the lines of why defvar and defcustom don't overwrite user settings. 

I see your point but the added complexity for this feature does not worth it IMO.
If you want to do it then add `def` prefix to the package name, something like `define-keymap-binding`.

**Comment by syl20bnr** [comment]
The `def` things is not even consistent in the Emacs API :-(
Anyway, this is not how key bindings works in Emacs and I fear about the added complexity for something that feels out of the Emacs way.

**Comment by syl20bnr** [comment]
This could be modeled after `setq` with a `bind-keymap-default`.

**Comment by justbur** [comment]
The ability to use default bindings is already there. It's entirely up to
you if you want to use it. It's not a big deal.

**Comment by justbur** [comment]
I think you're misunderstanding. See the currently named
olig-set-key-defaults function. That's all I'm talking about.

**Comment by syl20bnr** [comment]
I see, I took the problem the other way around it seems ;-)
I've been messed up by the flycheck issue which indeed is not related to this. Just to make my point clear about flycheck issue: this is a non issue for me so don't add complexity to attempt to fix it (i.e. make the binding available only when flycheck is enables).

**Comment by justbur** [comment]
I see. To be honest I just glanced at it so I don't really know what the
issue is with flycheck :-)

**Comment by TheBB** [comment]
Okay I'll explain the flycheck issue.

Some of the bindings under `SPC e` currently only makes sense when flycheck is enabled. When it is not enabled they don't do what they should (predictably). As far as I can tell they do not error out, but that's obviously not something we can rely on for all minor mode functions we might want to have in the leader keymap.

IMO the cleanest way would be to have bindings like `SPC e c` only be available when `flycheck-mode` is on and not when it's off. In practice this is precisely the same as the minor mode keymap except it's in the leader keymap, which, as I understand it, cannot be done with `evil-leader`.

The whole discussion about autoloading has clouded the issue. It's not about that.

**Comment by justbur** [comment]
Ok. The way that could be handled with this package is to have a flycheck
"leader map" that is bound using flycheck-mode-map for the value of
:mode-maps and to bind it to SPC ec.

**Comment by syl20bnr** [comment]
OMG I'm lost, let me re-read one more time :-D

**Comment by syl20bnr** [comment]
Ok I see, what's the difference with binding it directly in `flycheck-mode-map` ?

**Comment by justbur** [comment]
That's all this does but you have to bind multiple times for each key you
want to activate the map. It's pretty simple. There's no magic.

**Comment by syl20bnr** [comment]
I guess this too much implementation specific, what does it look like from a client point of view ?

**Comment by TheBB** [comment]
> Ok I see, what's the difference with binding it directly in flycheck-mode-map ?

I had to check. If I do

    (define-key flycheck-mode-map (kbd "SPC e q") 'hello)

it shows up in `which-key` but I still get `SPC e q is undefined` if I try to access it.

**Comment by justbur** [comment]
You want to put that in the correct evil map

**Comment by syl20bnr** [comment]
> Ok. The way that could be handled with this package is to have a flycheck "leader map" that is bound using flycheck-mode-map for the value of :mode-maps and to bind it to SPC ec.

This is done in olig hook hooked to flycheck that's it ?
But then if the user bound something directly in `flycheck-mode-map` under the same key do we overwrite her key binding ? I don't know if it is a valid point but maybe we will confuse the user by lazy-setting key bindings.

**Comment by syl20bnr** [comment]
Pweeeeee this is giving me headache.
To be clear, one of the feature I like from `evil-leader` is that it does not require the package to be loaded, everything depends on alist of symbols.

**Comment by TheBB** [comment]
I don't really understand too much about key bindings. I'm satisfied if it's understood that this is a feature that I would like (and which I—strictly from the point of view as a user—really like from evil.)

**Comment by syl20bnr** [comment]
The feature is `being able to bind keys for major modes _and_ minor modes`, am I correct ? :-)

**Comment by TheBB** [comment]
Yes.

**Comment by justbur** [comment]
> To be clear, one of the feature I like from evil-leader is that it does not require the package to be loaded, everything depends on alist of symbols and no keymap needs to be defined at the evil-leader/set-key call time.

If this is an important feature then I can use mode symbols to load the key maps. That's not a big problem.

**Comment by justbur** [comment]
Ok, new version that uses symbols and doesn't require anything to be loaded. Here's a working example
```elisp
  (defun test1 () (interactive) (message "1"))
  (defun test2 () (interactive) (message "2"))
  (bind-map base-leader-map
    :keys ("M-n")
    :evil-keys ("!"))
  (define-key base-leader-map "o" 'test1)
  (bind-map emacs-leader-map
    :keys ("M-n m")
    :evil-keys ("! !")
    :prefix-cmd emacs-prefix
    :major-modes (emacs-lisp-mode)
    :minor-modes (flycheck-mode))
  (define-key emacs-leader-map "t" 'test2)
```

major and minor modes are supported. The map can be defined already, but if it's not the macro will generate a new one.

**Comment by syl20bnr** [comment]
The interface looks good. What is `prefix-cmd` ?

Le dimanche 15 novembre 2015, Justin Burkett <notifications@github.com> a
écrit :

> Ok, new version that uses symbols and doesn't require anything to be
> loaded. Here's a working example
>
>   (defun test1 () (interactive) (message "1"))
>   (defun test2 () (interactive) (message "2"))
>   (bind-map base-leader-map
>     :keys ("M-n")
>     :evil-keys ("!"))
>   (define-key base-leader-map "o" 'test1)
>   (bind-map emacs-leader-map
>     :keys ("M-n m")
>     :evil-keys ("! !")
>     :prefix-cmd emacs-prefix
>     :major-modes (emacs-lisp-mode)
>     :minor-modes (flycheck-mode))
>   (define-key emacs-leader-map "t" 'test2)
>
> major and minor modes are supported. The map can be defined already, but
> if it's not the macro will generate a new one.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/justbur/olig/issues/1#issuecomment-156819578>.
>


-- 
-syl20bnr-

**Comment by justbur** [comment]
Just a name to use for declare-prefix-command if you want it.
